src/state/connection/hooks.ts

```diff
@@ -14,11 +15,35 @@ export {
   useTokenBalancesWithLoadingIndicator,
 } from 'lib/hooks/useCurrencyBalance'
 
+export type BalancesResult = { [tokenAddress: string]: CurrencyAmount<Token> | undefined }
+export type PricesResult = { [tokenAddress: string]: number | undefined } | undefined
+
 // mimics useAllBalances
-export function useAllTokenBalances(): [{ [tokenAddress: string]: CurrencyAmount<Token> | undefined }, boolean] {
+export function useAllTokenBalances(): [BalancesResult, boolean] {
   const { account } = useWeb3React()
   const allTokens = useDefaultActiveTokens()
   const allTokensArray = useMemo(() => Object.values(allTokens ?? {}), [allTokens])
   const [balances, balancesIsLoading] = useTokenBalancesWithLoadingIndicator(account ?? undefined, allTokensArray)
   return [balances ?? {}, balancesIsLoading]
 }
+
+export function useTokenPrices(balances: BalancesResult): PricesResult {
+  const [prices, setPrices] = useState<PricesResult | undefined>(undefined)
+  useEffect(() => {
+    loadPrices()
+    async function loadPrices() {
+      const tokensWithBalance: CurrencyAmount<Token>[] = Object.values(balances).filter(
+        (balance) => balance?.greaterThan(0) // filters undefined values so we can cast safely
+      ) as CurrencyAmount<Token>[]
+      const pricesEntries = await Promise.all(
+        tokensWithBalance.map(async (token) => {
+          const price = await fetchUSDPrice(token)
+          return [token.currency.address, price]
+        })
+      )
```

Comment:
This feels sketchy to me. Can you take a look at the way @cartcrom accomplished this in the mini portfolio? I think you'll be able to fetch a user's value-ordered balances using a single query instead of `Promise.all`'ing a potentially very large set of single token price requests.